### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by demberto** [construct/construct#994]
I recently replaced all binary structures with equivalent (or better) `construct.Struct`s in my library [PyFLP](https://github.com/demberto/PyFLP).

Earlier, I had my own weird implementation where I had a wrapper over `io.BytesIO` which allowed me to read C types from a stream directly. It utilised *also my library* [bytesioex](https://github.com/demberto/bytesioex). With my implementation, most structs were really lazily parsed and looked like this:

> ```py
> class Struct:
>     PROPS = {
>         "a": "I",
>         "b": "H"
>     }
> ```
>
> `PROPS` is a dict containing a mapping of field names like how they are in `construct.Struct` and the values are `struct` format specifiers.
>
> Values were directly read off of the stream, converted to Python types via the stdlib `struct` module in `bytesioex`.
---
The tests took around **0.8** seconds with my implementation. When I switched over to `construct.Struct` completely, it took a whooping **3** seconds or higher. However the compile feature helped it bring down to **2.2** seconds. Also I didn't have success to say, with `LazyStruct` #989.

The slowest part to parse is [this struct](https://github.com/demberto/PyFLP/blob/b3999df0f7c71805eb86bdfde51a4bbc5274302f/pyflp/arrangement.py#L124-L142), there are a total of 1000 of these struct instances in my test data. Due to the undocumented nature of these structs, using `Optional` is a must.

If you might have any suggestions, I would honestly like to know.

**Comment by arekbulski** [comment]
I was going to suggest the compilation feature but then, I see you already tested that. Have you tried using PyPy? LazyStruct was implemented for hacking contests where reading each byte is slooow so the Lazy part of LazyStruct is that it avoids reading bytes from the stream as much as possible, not that it avoids computation. 

I am looking into your project.

**Comment by arekbulski** [comment]
I am not sure if Optional(GreedyBytes) is any better than just GreedyBytes. But this is hardly an optimization.

**Comment by demberto** [comment]
I mean still, whatever causes it to be like 2-3x slower if its finally just reading and writing to a stream? 

> Have you tried using PyPy?

I will definitely try it.

A small question, are `FormatField`s compiled versions (aka `struct.Struct` instances or are just format specifiers used directly every time)?

**Comment by arekbulski** [comment]
Why is it slower than just struct module and io.BytesIO? Because Constract parsing classes are wrappers around wrappers around basic operations like struct.unpack. Look at the compilation page on docs site, I laid out there some ideas why compiled version work faster.

You can use APT to install pypy:  `sudo apt install pypy pypy3`  etc.

What are you asking exactly, about FormatField-s?

**Comment by arekbulski** [comment]
Look at the source if you want to know how it works:
https://github.com/construct/construct/blob/2626b398a676191114f6ce5d8db903e9590700db/construct/core.py#L982

**Comment by demberto** [comment]
Okay, thanks I am gonna experiment tomorrow with PyPy and let you know the results

**Comment by arekbulski** [comment]
Maybe we could reimplement a StructOfOptionals parsing class that would be more direct than a Struct of Optionals (which are really Select of Pass)?

**Comment by demberto** [comment]
Yea that would be a pretty niche use case imo (speaking generally). I think its not the wrappers itself which cause all the delay, its some magic going on in the code itself. Maybe some if it is caused by the dependency on `construct_typed`

**Comment by arekbulski** [comment]
There is no magic fairy dust sprinkled over it, it is the thin wrappers around wrappers that cause the useless computation. I am willing to help you out with writing that SoO parsing class but I need to know what are the parsing semantics are to be. What if SoO encounters EOF mid sequence. Does it need to return None for all following subcons? Can it skip them?

**Comment by demberto** [comment]
> I am willing to help you out with writing that SoO parsing class

That's really very kind of you, it will definitely make the code cleaner. But honestly it would be a niche use case for general purpose declarative parsing. Plus, would it really have any performance boost?

> What if SoO encounters EOF mid sequence?

The struct size is not fixed, that's the primary reason I am using `Optional` fields. In this way, I can atleast get the values for the fields parsed from whatever data there was, while not having to make different structs for every possible struct change.

> Does it need to return None for all following subcons?

Yes, If the stream wasn't big enough such that some subcons didn't have data to parse they should return `None`,  just like `Optional` fields do.

> Can it skip them?

If there's no more data in the stream, what is the benefit of not skipping them anyways?

**Comment by arekbulski** [comment]
The list of parsing classes that were added to the core library that were requested by 1-3 people is as long as my arm. For example, LazyStruct and LazyArray that you tried were made for one person, period. GreedyBytesEx (now called OffsetedEnd) was made for like 2 people I think. Symetric Encryption classes were made by one person for probably just himself (Timmy made it, the guy behind construct-typing). Need I go on? I wrote in the documentation page at the very top "Feel free to request any changes that would support your project." Yes?

You were the one complaining about performance so I ask you, do you want to drill this hole or not? 

Its a basic question about SoO behavior. Does it need to return None for remaining subcons (like Struct of Optionals does) or can it return a Container with half of keys missing? This is a simple question. Skipping subcons will have a performance impact.

**Comment by demberto** [comment]
Oh okay, thanks :))

> Does it need to return None for remaining subcons (like Struct of Optionals does)

Either one is okay really. I don't use the constructs directly. There's a wrapper (yea, one more 😄) over them in my library. If construct returns None I can return None directly and if the key is not found I can return None myself or maybe reraise the exception.

I am debating myself whether I should return None when accessing attributes which weren't parsed, or I should raise an exception. This is for the getter, the setter raises an exception if it doesn't find the key or it is set to None, maybe you can provide some insight into this?

**Comment by arekbulski** [comment]
> Btw would it possible to add a key later in this SoO if it wasn't parsed? That would lead to expansion of the stream, which is not a good idea but can be useful in the future.

I do not follow what you mean. Please explain in simple terms, or with an example.

**Comment by arekbulski** [comment]
> maybe you can provide some insight into this?

I am not sure what you are asking there either. Please elaborate?

**Comment by demberto** [comment]
> I do not follow what you mean. Please explain in simple terms, or with an example.

I just meant to say that `io.BytesIO` is expandable, unlike a `memoyview` (which is readable / writable but not expandable). I don't need an expandable stream, a `memoryview` should suffice if you choose to implement using it. This example precisely covers it:

```py
>>> import construct as c
>>> MyStruct = c.Struct(
...     "a" / c.Optional(c.Int32ul),
... )
>>> mystruct = MyStruct.parse(bytes(0))
>>> mystruct
Container(a=None)
>>> mystruct["a"] = 0
>>> MyStruct.build(mystruct)
b'\x00\x00\x00\x00'    # This shouldn't happen
>>> mystruct["a"] = None
>>> MyStruct.build(mystruct)
b''
```

In this format I am working on, dumping a struct bigger than the one which came in for parsing can cause version compatibility issues.

Now, for the second part of my initial reply, where I ask you for your insight, I have this logic already built-in in to my wrapper over the constructs.

```py
def __init__(self, id: int, data: bytes):
        super().__init__(id, data)
        self._struct = self.STRUCT.parse(data)

def __bytes__(self):
    self._data = self.STRUCT.build(self._struct)
    return super().__bytes__()

def __contains__(self, prop: str):
    return prop in self._struct

def __getitem__(self, key: str):
    return self._struct[key]     # Your view: Should this return None if key is not found i.e. parsed or raise an Exception?

def __setitem__(self, key: str, value: Any):
    if key not in self or self[key] is None:   # This covers either of the implementations you suggest
        raise PropertyCannotBeSet
    self._struct[key] = value
```

**Comment by arekbulski** [comment]
> b'\x00\x00\x00\x00'    # This shouldn't happen

This is correct behavior. Optional is actually a macro over Select(sc, Pass). Select tries to use the integer value when building Int32ul, building succeeds. There is nothing wrong there.

**Comment by demberto** [comment]
> There is nothing wrong there.

No, I didn't mean anything is **wrong**. You asked me on the semantics of parsing an SoO and I replied in context to that.

**Comment by arekbulski** [comment]
Okie dokie, it seems we are now on the same page. I will implement SoO for you tomorrow. Too tired atm.

**Comment by demberto** [comment]
Just tried Pypy 3.7-3.9. It takes anywhere between 15-30 seconds, as opposed to 3-5 seconds in CPython. But that's in pytest

**Comment by skarnproject** [comment]
I can also relate to the mentioned performance concerns. The gain we get from PyPy is actually irrelevant to the Construct library itself, it is just simple a faster interpret than CPython.

I've had a look at the code compilation feature and identified a few issues that also make the compiled code have non-optimal performance.

Construct currently draws no distinction between structs that use dynamic features (such as dynamic arrays, conditionals, repeaters, etc) and simplistic C-style statically type structures with known size and types.

```py
Struct(
    "num8" / Int8ub,
    "num8_1" / Int8ub,
)

```
Let's consider this structure. According to the compilation example, that will result into roughly this code when compiled:

```py
        this['num8'] = unpack('>B', read_bytes(io, 1))[0]
        this['num8_1'] = unpack('>B', read_bytes(io, 1))[0]
```

When struct.unpack is used like this for individual integers, on each invocation it has to be compile or look up a cache format string. This is done on every parsing of this structure, and every building.

Instead, this could have been:
`struct.Struct('>BB')
`
That is compiled once and cached once per the structure type. This is the fastest IO solution there is for pure Python. I measured ctypes.Struct (about twice slower) as well as individual struct.unpack / struct.pack calls (many times slower). int.from_bytes is a little faster, but the speedup is only relevant if we read / write a single integer.

The fastest solution for reading arrays in default Python that I found is:
For builtin-type singular element arrays: Python's array module.
For arrays of structures: a struct.Struct containing a format string with repeated instances of the structure.
Consider an array of structures from the previous example:
`struct.Struct('>BB' * 20)`

This works faster than reading the same structure 20 times in a row into a list.

It would be good to be able to identify static blocks within the compound structure and create a parsing and building tree that would have those static parts at optimized and pre-compiled as possible. Ideally we only need to break the static format string in the following cases: 
1) change of endianness. Python's struct expects endianness to be the first '>' or '<' of the format string.
2) A non-standard byte size integer. Such as 24-bit integer.
3) Dynamic array, conditional or any other dynamic feature.

Once such as system is implemented, another potential performance improvement can be achieved with compilation into Cython code. I saw on the doc page that there was priorly such a feature which yielded bad performance results.
Considering the example on the page, it is clear that it did not use Cython's static typing extensively enough.

A parsing (building) tree of static and dynamic blocks that we presume to help, can allows us to generate C structures for static blocks for the Cython output, avoiding the need to use Python's struct module.

`struct.Struct('>BB' * 20)`

can become:

```
cdef struct StructName:
    int8 num8
    int8 num8_1
```

This should yield performance close to C when reading these structures from memory.

This sounds like a pretty extensive change. I've seen on the docs that there is a total rewrite of construct planned. Is there a chance the performance issue will be taken care of there? If so, is there any way to contribute to these efforts at the moment?

**Comment by demberto** [comment]
@skarnproject Nice work there.

I think using a `memoryview` (with a pointer to current location, optionally) can yield better performance than using a `io.BytesIO` in cases where the size is predetermined and is not going to change in the lifetime of a struct instance.

I recently came across [f-enum](https://pypi.org/project/f-enum/) which gave me a 50% speed boost (literally for free). My code uses enums a lot, so YMMV. Enums are commonly used in structs so it was worth letting you know.

**Comment by arekbulski** [comment]
> I've seen on the docs that there is a total rewrite of construct planned. 

Incorrect. Total rewrite has  (in effect) already happened. Not going to happen. Can you please point me to it?
